dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java

```diff
@@ -288,10 +288,11 @@
             putMsg(result, Status.TASK_DEFINE_NOT_EXIST, taskCode);
             return result;
         }
-        TaskDefinitionLog taskDefinitionLog = taskDefinitionLogMapper.queryByDefinitionCodeAndVersion(taskCode, version);
-        taskDefinitionLog.setUserId(loginUser.getId());
-        taskDefinitionLog.setUpdateTime(new Date());
-        int switchVersion = taskDefinitionMapper.updateById(taskDefinitionLog);
+        TaskDefinitionLog taskDefinitionUpdate = taskDefinitionLogMapper.queryByDefinitionCodeAndVersion(taskCode, version);
+        taskDefinitionUpdate.setUserId(loginUser.getId());
+        taskDefinitionUpdate.setUpdateTime(new Date());
+        taskDefinitionUpdate.setId(taskDefinition.getId());
```

Comment:
I do not think directly set `task definitions log` id here. I think both` task definition` and `task definition log` id could be auto increase. Am I wrong @JinyLeeChina

Reply:
in my opinion ,  switchVersion is to set  the version  of   t_ds_task_definition  , which reference from   t_ds_task_definition_log,   let's see the metadata,  version=6  id=8  in t_ds_task_definition, but version=6 id=22 in t_ds_task_definition_log,   taskDefinitionMapper.updateById(t_ds_task_definition_log)  will  throw "Switch task definition version error" because  id=22 not in t_ds_task_definition,  or maybe I am wrong, but the problem I got when  I call  openapi  "switchTaskDefinitionVersion" 
![image](https://user-images.githubusercontent.com/13324943/141246118-994d61aa-e3d4-45ae-b17e-903c55fa6b38.png)

Reply:
I go to our UI try to find out which place call service `switchVersion` but failed 😢 Did one know where we call it?

Reply:
Oh, I got your point. You means that function switchVersion it's get task definition from task_definition_log and recover the snapshot to task_definition? If so your change make sense.

But the total logic is odd, could anyone tell me where would call this function?

Reply:
I think it's not used yet, because task versions is not open to users now ,   I'am just testing all openapis by create httprequest to help myself get familiar with the code

Reply:
> I do not think directly set `task definitions log` id here. I think both` task definition` and `task definition log` id could be auto increase. Am I wrong @JinyLeeChina

Yes, you're right. In `t_ds_task_definition` and `t_ds_task_definition_log`, `Id` has no business meaning, just as a table to auto-increase the primary key. During `switchTaskDefinitionVersion` , the `update_time` and `id` of the main table data and log table will be different, and others will be the same

Reply:
update t_ds_task_definition by  taskDefinitionLog  will never succeed.   int switchVersion = taskDefinitionMapper.updateById(taskDefinitionLog);  line 288 , TaskDefinitionServiceImpl.java  id is differnet between t_ds_task_definition and t_ds_task_definition,  you may see the metadata of picture and it's explanation above

Reply:
One line of code is missing here, `taskDefinitionLog.setId(taskDefinition.getId());` , Can you submit a PR to fix it ？

Reply:
oh, I was careless, you have added.

Reply:
this pr is to fix what you mentioned ,   taskDefinitionUpdate.setId(taskDefinition.getId());     (line 294)              and  I just change  the java obj name "taskDefinitionLog" to  "taskDefinitionUpdate"  is to underline that we get obj from t_ds_task_definition_log  it to update t_ds_task_definition

Reply:
right, very good